### DISCUSSION TITLE
New version and ideas gma3 lib

### DISCUSSION THREAD

**Discussion by SkowroFl** [sstaub/gma3#6] [Ideas]
Hello Stefan,

are you planning a newer Version V3 of your gma3 library? I have some ideas to create more flexibilty with the code:

- Would it be possible to implement a function which allows you to change the pages of a key or function during the loop? e.g. by Key.update(PageX);
- On my board I am using illuminated encoders with a button. This button is connected to +5V, so it is not possible to use it with the internal pullup of the controller. Maybe a new class for encoder button which fires a macro? Only difference no INPUT_PULLUP
- My code uses a lot of SRAM, because I have a lot of keys and faders. Have you an idea to reduce that?

Thanks a lot,
Florian

**Comment by sstaub** [comment]
Hi Florian,
- a class for paging is a good idea, maybe two button for up and down
- you can modify the cmdButton class, no Pullup, but you have to reverse the execution code inside of update
- the use of SRAM depends on the board type and compiler, which board did you use? What say the IDE of RAM usage?

**Reply by SkowroFl** [reply]
Hey Stefan,

- very nice, I am looking forward to that
- I found the part in the code, maybe this could be interested for somebody who is a beginer
- on my breadboard is actualy an arduino mega 2560 r3, the IDE tells that global variables are using 8641 Bytes (105%) of dynamic storage (by full using of my application, maybe I have to reduce the functions or change the board, 10 faders, 20 encoders, 40 buttons, 20 cmd :D)

**Reply by sstaub** [reply]
- Can you try v1? It can only UDP and you need the OSC CNMAT library (have a look to the readme for that version).
I think it need less RAM, but the performance quality can be a problem.
- A general paging will cause some problems with the faders when they are not motorized, I have to think about.

**Reply by sstaub** [reply]
you can try adjust following  sizes 
```
// OSC settings
#define OSC_PATTERN_SIZE 64
#define OSC_STRING_SIZE  64
#define OSC_MESSAGE_SIZE 256
```
Because I use no dynamic memory to avoid leaks
Try following:
```
// OSC settings line 30 in gma3.h
#define OSC_PATTERN_SIZE 48 // this for the OSC pattern, length depends on naming conventions
#define OSC_STRING_SIZE  32 // this is for the command line
#define OSC_MESSAGE_SIZE 96 // this is the complete message size
```

**Comment by sstaub** [comment]
Made a new version with updated allocation, see also the readme for optimization.

**Reply by SkowroFl** [reply]
With the optimazation of the sizes you can decrease the memory (48/32/96). In my case from 8641 bytes to 6401 bytes. That is great. 
In my case I am using motorized faders, the swap from one page to another could I solve here with code.

**Comment by sstaub** [comment]
In the moment I'm working on a large library for ETC EOS consoles, but I will dig inside in holidays. Did you have a useful library for motorized fader? I have an ALPS Motorfader and a motor driver card, but didn't find the time to experiment with it.

**Reply by SkowroFl** [reply]
I am not shure if there exists a library for motorfader. In the internet are some usefull examples which I used to start coding (StartHardware- Youtube). At the moment I realize it with a 2 dimensional array [Page][Fadervalue], when receiving the string "Next page" it changes the row and starts to drive the faders. During this cycle no OSC is updated. At the moment it looks very promising, I am very satisfied with the result. This is just trying it out till the gma3 OSC library supports page swapping. Hardware is Alps RS60N11M9-LIN5K and the L293D.

**Comment by sstaub** [comment]
for motorfader I found following:
https://tttapa.github.io/Pages/Arduino/Control-Theory/Motor-Fader/
https://github.com/stagehacks/FADER_LIB_MIDI_FEEDBACK

I think about following concept
- paging buttons with two callback were you can insert your necessary code
- .page() function where you can change the page number on fly for each element,
this gives backwards compatibility and less RAM consumption  but the performant will decrease
using the same update functions like in v1

Over holidays I will make a new development branch you can use for testing
In the moment I'm writing a library for Sexton displays which can used for virtual elements (this is on my TODO list)

**Reply by SkowroFl** [reply]
No problem, take the time you need for that.

**Comment by sstaub** [comment]
Here is a new beta for testing. You can now change page global or local. Also a helper class for up/down button.
Have a look to the README and the gma3_Paging example.
Let me now if you find bugs or if there are performance problems, there should less RAM usage.
[gma3.zip](https://github.com/sstaub/gma3/files/7780079/gma3.zip)

**Comment by sstaub** [comment]
I'm preparing a new version with two new functions for faders
- value() which gives you the last fader value
- fetch(value) which locks sending OSC data until the value is reached
This will make fader handling very easy.
Q: Did you find a way to get the actual fader value from Console/onPC?

**Comment by SkowroFl** [comment]
Thanks for version V3, the change of pages works fine. At the moment I am just using the global change. I can agree that the RAM usage is much lower than in V2. Really awesome your work.

No, I do not found a way to send the actual fader values from console/onPC to AVR. In a 2- dimensional array are the fader values stored, when changing the pages I go up oder down in the array. So it is complete "standalone" with no data exchange. But better than nothing, maybe later someone founds a way to transfer data. In MA2 the only way was with the Webremote (RD Showcockpit), but in MA3 it works not yet.

**Comment by sstaub** [comment]
Found a way to get the fader value from the console:
- send an osc string manually and add 0.001 to the actual value
```
oscMessage("/gma3/Page1/Fader201", fader201.value() + 0.001f);
sendOSC();
```
this returns e.g.
`OSC Pattern: /gma3/13.13.1.5.2 String: FaderMaster Integer 1: 1 Integer 2: 0 Float: 14.00`
My problem in the moment is to decode 13.13.1.5.2
I will send you a new beta later.

**Comment by sstaub** [comment]
Here is a new beta, new:
- fetch
- lock
- value
refer to readme
[gma3.zip](https://github.com/sstaub/gma3/files/7795788/gma3.zip)

**Comment by sstaub** [comment]
I forgot, the Selection class is replaced by a general Button class, see example and readme.
Next stepp is a fader jitter() function to make manual fetching easier,
and a console() function for getting the fader value from console.

**Comment by sstaub** [comment]
This is the release candidate. In the moment I didn't found a way to get the console data.
The problem that the Echo Data from console include no information about the executor assignment.
The Echo in the example before `13.13.1.5.2` is about sequence list 2 (which is assigned to page 1 / exec 201).
So the Echo is very cryptic and have nothing to do with the assignment of the executor elements.
[gma3rc.zip](https://github.com/sstaub/gma3/files/7798214/gma3rc.zip)
